File: apps/server/src/application/pilot/handlers/create-pilot-product.handler.ts
import { type Effect, gen } from 'effect/Effect'

import {
  makeNotSynced,
  makePilotProduct,
  makePilotProductPublished,
  type PilotProductCreationError,
  requiresChangeNotification,
} from '../../../domain/pilot'
import { Clock, EventPublisher, IdGenerator, PilotProductRepository } from '../../../ports/driven'
import { type ValidatedProductData, validateProductData } from '../validation'
import { createVariants } from '../mappers/variant.mapper'
import { publishEvent } from '../../shared/event-helpers'
import type { PilotProduct } from '../../../domain/pilot'
import type { PilotProductCreationCommand } from '../commands'

// #region HANDLER: CREATE PILOT PRODUCT

export const pilotProductCreationHandler = (
  command: PilotProductCreationCommand
): Effect<
  PilotProduct,
  PilotProductCreationError,
  PilotProductRepository | IdGenerator | EventPublisher | Clock
> =>
  gen(function* () {
    const validated = yield* validateProductData(command.data)

    const product = yield* createAggregate(validated)

    const repo = yield* PilotProductRepository
    const savedProduct = yield* repo.save(product)

    if (requiresChangeNotification(savedProduct)) {
      yield* emitEvent(savedProduct, command)
    }

    return savedProduct
  })

// #endregion

// #region AGGREGATE CREATION

const createAggregate = (
  validated: ValidatedProductData
): Effect<PilotProduct, never, IdGenerator | Clock> =>
  gen(function* () {
    const idGen = yield* IdGenerator
    const clock = yield* Clock

    const productId = yield* idGen.generateProductId()
    const now = yield* clock.now()
    const variants = createVariants(validated.variants)

    return makePilotProduct({
      syncStatus: makeNotSynced(),
      id: productId,
      label: validated.label,
      type: validated.type,
      category: validated.category,
      description: validated.description,
      priceRange: validated.priceRange,
      variants,
      views: validated.views,
      status: validated.status,
      createdAt: now,
      updatedAt: now,
    })
  })

// #endregion

// #region EVENT EMISSION

const emitEvent = (
  product: PilotProduct,
  command: PilotProductCreationCommand
): Effect<void, never, EventPublisher | Clock> =>
  gen(function* () {
    const clock = yield* Clock
    const now = yield* clock.now()

    const event = makePilotProductPublished({
      productId: product.id,
      product,
      correlationId: command.correlationId,
      userId: command.userId,
      timestamp: now,
    })

    yield* publishEvent(event)
  })

// #endregion

File: apps/server-java/src/main/java/com/maisonamane/application/pilot/handler/CreatePilotProductHandler.java
package com.maisonamane.application.pilot.handler;

import com.maisonamane.application.pilot.command.CreatePilotProductCommand;
import com.maisonamane.application.pilot.validation.ProductDataValidator;
import com.maisonamane.application.pilot.validation.ValidatedProductData;
import com.maisonamane.domain.pilot.aggregate.PilotProduct;
import com.maisonamane.domain.pilot.enums.ProductStatus;
import com.maisonamane.domain.pilot.error.PilotProductError;
import com.maisonamane.domain.pilot.error.PilotProductError.PersistenceError;
import com.maisonamane.domain.pilot.event.PilotProductPublished;
import com.maisonamane.port.driven.repository.PilotProductRepository;
import com.maisonamane.port.driven.service.Clock;
import com.maisonamane.port.driven.service.EventPublisher;
import com.maisonamane.port.driven.service.IdGenerator;
import io.vavr.control.Either;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

/**
 * Handler for CreatePilotProductCommand.
 * Maps Effect-TS Effect.gen handler to Java @Service with flatMap chains.
 *
 * Effect.Effect<T, E, R> → Either<E, T> with constructor-injected dependencies
 * yield* sequencing → .flatMap() chains
 */
@Service
public class CreatePilotProductHandler {

    private static final Logger log = LoggerFactory.getLogger(CreatePilotProductHandler.class);

    private final PilotProductRepository repository;
    private final IdGenerator idGenerator;
    private final Clock clock;
    private final EventPublisher eventPublisher;

    public CreatePilotProductHandler(
            PilotProductRepository repository,
            IdGenerator idGenerator,
            Clock clock,
            EventPublisher eventPublisher) {
        this.repository = repository;
        this.idGenerator = idGenerator;
        this.clock = clock;
        this.eventPublisher = eventPublisher;
    }

    /**
     * Handle the create pilot product command.
     *
     * Railway-oriented programming with Either:
     * 1. Validate product data
     * 2. Create aggregate
     * 3. Save to repository
     * 4. Publish event if PUBLISHED
     */
    public Either<PilotProductError, PilotProduct> handle(CreatePilotProductCommand command) {
        return ProductDataValidator.validate(command.data())
                .mapLeft(err -> (PilotProductError) err)
                .flatMap(this::createAggregate)
                .flatMap(this::saveProduct)
                .peek(product -> {
                    if (product.status() == ProductStatus.PUBLISHED) {
                        emitEvent(product, command);
                    }
                });
    }

    private Either<PilotProductError, PilotProduct> createAggregate(ValidatedProductData validated) {
        var productId = idGenerator.generateProductId();
        var now = clock.now();

        var product = PilotProduct.create(
                productId,
                validated.label(),
                validated.type(),
                validated.category(),
                validated.description(),
                validated.priceRange(),
                validated.variants(),
                validated.views(),
                validated.status(),
                now);

        return Either.right(product);
    }

    private Either<PilotProductError, PilotProduct> saveProduct(PilotProduct product) {
        return repository.save(product)
                .mapLeft(err -> PersistenceError.of(err.message(), err.cause()));
    }

    private void emitEvent(PilotProduct product, CreatePilotProductCommand command) {
        var now = clock.now();
        var event = PilotProductPublished.of(
                product,
                command.correlationId(),
                command.userId(),
                now);

        eventPublisher.publish(event)
                .peekLeft(error -> log.error(
                        "Failed to publish event, will be retried. Error: {}",
                        error.message()));
    }
}
